### ISSUE TITLE
Bug: `non-nullable-type-assertion-style` false positive for generic functions

### ISSUE THREAD

**Issue by atitcreate** [oxc-project/tsgolint#954]
cd to a temp dir, run:
```bash
npm install --save-dev eslint typescript typescript-eslint \
    oxlint oxlint-tsgolint
npx tsc --init
```
create `eslint.config.mjs`:
```js
import { defineConfig } from "eslint/config"
import tseslint from "typescript-eslint"
export default defineConfig({
  files: ["test.ts"],
  extends: [tseslint.configs.base],
  rules: {
    "@typescript-eslint/non-nullable-type-assertion-style": "error",
  },
  languageOptions: {
    parserOptions: {
      projectService: true,
    },
  },
})
```
create `.oxlintrc.json`:
```jsonc
{
  "options": { "typeAware": true },
  "categories": { "correctness": "off" },
  "rules": {
    "typescript/non-nullable-type-assertion-style": "error"
  }
}
```
create `test.ts`:
```ts
const href = (document.querySelector(".anchor") as HTMLAnchorElement).href
```

```bash
npx eslint ; npx oxlint
# Both prompt "Use a ! assertion to more succinctly remove null and undefined from the type."
```

Both `npx eslint --fix` and `npx oxlint --fix` apply the same fix:
```patch
-const href = (document.querySelector(".anchor") as HTMLAnchorElement).href
+const href = (document.querySelector(".anchor")!).href
```

The issue is in typescript-eslint repo labelled "external" "wontfix" "working as intended"
> https://github.com/typescript-eslint/typescript-eslint/issues/528#issuecomment-493140171
> https://github.com/typescript-eslint/typescript-eslint/issues/9007#issuecomment-2080161404

Maybe at least for oxlint the auto-fix can be marked as "dangerous fix"?
